util/util.go

```diff
@@ -128,24 +115,72 @@ func GetServiceIDByPostmanCollectionID(c rest.FrontClient, collectionID string)
 
 		if strings.EqualFold(collectionID, svcCollectionID) {
 			result = svc.ID
-			postmanCollectionIDCache.Set(svcCollectionID, svc.ID, cache.DefaultExpiration)
 		}
 	}
 
-	if (result != akid.ServiceID{}) {
-		printer.Stderr.Debugf("Postman collectionID %q is %q\n", collectionID, result)
-		return result, nil
+	return result, nil
+}
+
+func GetOrCreateServiceIDByPostmanCollectionID(c rest.FrontClient, collectionID string) (akid.ServiceID, error) {
+	// Normalize the collectionID.
+	collectionID = strings.ToLower(collectionID)
+	unexpectedErrMsg := "Something went wrong while starting the Agent. " +
+		"Please contact Postman support (observability-support@postman.com) with the error details"
+	failedToCreateServiceErrMsg := "Failed to create service for given collectionID: %s\n"
+
+	if id, found := postmanCollectionIDCache.Get(collectionID); found {
+		printer.Stderr.Debugf("Cached collectionID %q is %q\n", collectionID, akid.String(id.(akid.ServiceID)))
+		return id.(akid.ServiceID), nil
+	}
+
+	// Fetch service and fill cache
+	ctx, cancel := context.WithTimeout(context.Background(), apiTimeout)
+	defer cancel()
+
+	serviceID, err := GetServiceIDByPostmanCollectionID(c, ctx, collectionID)
+	if err != nil {
+		printer.Stderr.Debugf("Failed to get list of services associated with the API Key: %s\n", err)
+		return akid.ServiceID{}, errors.Wrap(err, unexpectedErrMsg)
+	}
+
+	if (serviceID != akid.ServiceID{}) {
+		printer.Stderr.Debugf("ServiceID for Postman collectionID %q is %q\n", collectionID, serviceID)
+		postmanCollectionIDCache.Set(collectionID, serviceID, cache.DefaultExpiration)
+		return serviceID, nil
 	}
 
 	name := postmanRandomName()
 	printer.Debugf("Found no service for given collectionID: %s, creating a new service %q\n", collectionID, name)
 	// Create service for given postman collectionID
 	resp, err := c.CreateService(ctx, name, collectionID)
 	if err != nil {
-		printer.Stderr.Debugf("Failed to create service for given collectionID: %s\n", err)
+		httpErr, ok := err.(rest.HTTPError)
+		if !ok {
+			printer.Stderr.Debugf(failedToCreateServiceErrMsg, err)
+			return akid.ServiceID{}, errors.Wrap(err, unexpectedErrMsg)
+		}
+
+		var errorResponse rest.CreateServiceErrorResponse
+		if err := json.Unmarshal(httpErr.Body, &errorResponse); err != nil {
+			printer.Stderr.Debugf(failedToCreateServiceErrMsg, err)
+			return akid.ServiceID{}, errors.Wrap(err, unexpectedErrMsg)
+		}
 
-		if httpErr, ok := err.(rest.HTTPError); ok && httpErr.StatusCode == 403 {
-			error := fmt.Errorf("You cannot send traffic to the collection with ID %s. "+
+		if httpErr.StatusCode == 409 && errorResponse.Message == "collection_already_mapped" {
```

Comment:
Please add a comment explaining the use case which requires this (a race between multiple agents trying to create the mirrored service.)